### PR TITLE
Add `filter(func)` method

### PR DESCRIPTION
**Pull request by GlenHertz**

### CODE REVIEW

**Comment by GlenHertz** [comment]
I guess I'd need to add some tests.  But please let me know if this feature is acceptable.

**Comment by jakobnissen** [comment]
If this is to be done, I think it should be done consistently for `filter`, `Iterators.filter`, `map`, `Iterators.map`, `reduce`, `mapreduce`, `foldl` and `foldr`.

**Comment by simeonschaub** [comment]
> If this is to be done, I think it should be done consistently for filter, Iterators.filter, map, Iterators.map, reduce, mapreduce, foldl and foldr. 

See https://github.com/JuliaLang/julia/issues/35293

**Comment by JeffBezanson** [comment]
From triage: we can't do this for `map` because the 0-collection case is useful (this has been discussed before). However it can be done for all the other listed functions.

- For `mapreduce` I assume you'd need to pass two functions up front? The result should always be a 1-argument function that accepts a collection/iterator.
- reduce, foldl, and foldr also need to accept an `init` keyword arg (on the first call, IMO).
- Idea from @StefanKarpinski : add `mapper(f)` to handle the `x->map(f, x)` case.

File: base/array.jl [hunk 1]
```diff
@@ -2502,6 +2502,29 @@ function filter!(f, a::AbstractVector)
...
+
+# Example usage:
+```jldoctest
+julia> -5:5 |> filter(<(-3))
+2-element Vector{Int64}:
+ -5
+ -4
+```
+!!! compat "Julia 1.8"
+This functionality requires at least Julia 1.8
```
[the remark this answers is not in the document]
Comment (GlenHertz) [reply base/array.jl:2522]: Thanks.

File: base/array.jl [hunk 2]
```diff
@@ -2502,6 +2502,29 @@ function filter!(f, a::AbstractVector)
...
+"""
+    filter(f)
+
+Create a function that filters its arguments with function `f` using [`filter`](@ref), i.e.
+a function equivalent to `x -> filter(f, x)`.
+
+The returned function is of type `Base.Fix1{typeof(filter)}`, which can be
+used to implement specialized methods.
+
+# Example usage:
```
Comment (johnnychen94) [inline base/array.jl:2514]:
```suggestion
# Examples
```

File: base/array.jl [hunk 3]
```diff
@@ -2502,6 +2502,29 @@ function filter!(f, a::AbstractVector)
...
+```jldoctest
+julia> -5:5 |> filter(<(-3))
+2-element Vector{Int64}:
+ -5
+ -4
+```
+!!! compat "Julia 1.8"
+This method requires at least Julia 1.8.
```
Comment (johnnychen94) [inline base/array.jl:2522]:
```suggestion
!!! compat "Julia 1.8"
    This method requires at least Julia 1.8.
```

File: base/array.jl [hunk 4]
```diff
@@ -2502,6 +2502,29 @@ function filter!(f, a::AbstractVector)
...
+
+The returned function is of type `Base.Fix1{typeof(filter)}`, which can be
+used to implement specialized methods.
+
+# Example usage:
+```jldoctest
+julia> -5:5 |> filter(<(-3))
+2-element Vector{Int64}:
+ -5
+ -4
```
Comment (mcabbott) [inline base/array.jl:2519]: I presume that the motivation for this is to pass it to higher-order functions, so perhaps the examples should show that? 
```suggestion
julia> (1, 2, NaN, 4) |> filter(!isnan)
(1, 2, 4)

julia> map(filter(iseven), [1:3, 2:4, 3:5])
3-element Vector{Vector{Int64}}:
 [2]
 [2, 4]
 [4]
```

**Review by GlenHertz — COMMENTED** [review]
Nice!

**Comment by GlenHertz** [comment]
Is this good to go now?

**Comment by ViralBShah** [comment]
Bump. Merge?

**Review by JeffBezanson — APPROVED** [review]

File: base/array.jl [hunk 5]
```diff
@@ -2633,6 +2633,33 @@ function filter!(f, a::AbstractVector)
...
+(1, 2, 4, 6)
+
+julia> map(filter(iseven), [1:3, 2:4, 3:5])
+3-element Vector{Vector{Int64}}:
+ [2]
+ [2, 4]
+ [4]
+```
+!!! compat "Julia 1.8"
+    This method requires at least Julia 1.8.
```
Comment (mcabbott) [inline base/array.jl:2657]: Should be 1.9 right?

Comment (StefanKarpinski) [reply base/array.jl:2657]:
```suggestion
!!! compat "Julia 1.9"
    This method requires at least Julia 1.9.
```

Comment (StefanKarpinski) [reply base/array.jl:2657]: Was hoping that GitHub provides a way to make a new PR from a comment like this. Alas...

Comment (StefanKarpinski) [reply base/array.jl:2657]: Created the old fashioned way: https://github.com/JuliaLang/julia/pull/47045https://github.com/JuliaLang/julia/pull/47045.

**Pull request merged**